lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -998,6 +998,41 @@ SmallVector<unsigned> DotOperandEncodingAttr::getShapePerCTATile(
   }
 }
 
+LogicalResult DotOperandEncodingAttr::verify(
+    ::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError,
+    unsigned opIdx, Attribute parent, unsigned kWidth) {
+  if (opIdx != 0 && opIdx != 1)
+    return emitError() << "DotOperandEncoding::opIdx can be 0 or 1, got "
```

Comment:
The error will be printed in the context of IRs. So using IR mnemonic would be better. Like "dot_op attribute opIdx can only be ..".

Similarly for the following error messages.

Reply:
I don't think this function is related to IR printing.
It could be called at any time, including in the middle of optimization passes.
From my point it is better to give C++ name in the error message, so developer can find problematic place faster.

I can try to embed encoding dump in error, as a compromise, what do you think?

Reply:
> It could be called at any time, including in the middle of optimization passes.

You can call it in the middle of optimization passes when it errors out what you see is still error message pointing to a specific IR location with surrounding IRs? 

> From my point it is better to give C++ name in the error message, so developer can find problematic place faster.

Note that this is validation logic for the attribute. If we see errors on this attribute, it's likely somewhere upper layers is wrong when lowering to it. So we don't need to fix this attribute's C++ code per se? Unless we have wrong validation.

Reply:
> You can call it in the middle of optimization passes when it errors out what you see is still error message pointing to a specific IR location with surrounding IRs?

Verify method is called before every construction of attribute if asserts are enabled. In case this error appears somewhere inside no IR is printed, error looks like this:
```
<unknown>:0: error: DotOperandEncodingAttr::kWidth currently supports only 16 for WMMA parent

triton-opt: /root/.triton/llvm/llvm-3a831621-ubuntu-x64/include/mlir/IR/StorageUniquerSupport.h:180:
static ConcreteT mlir::detail::StorageUserBase<ConcreteT, BaseT, StorageT, UniquerT,
Traits>::get(mlir::MLIRContext*, Args&& ...) [with Args = {unsigned int, mlir::Attribute, unsigned int};
ConcreteT = mlir::triton::gpu::DotOperandEncodingAttr; BaseT = mlir::Attribute; StorageT =
mlir::triton::gpu::detail::DotOperandEncodingAttrStorage; UniquerT = mlir::detail::AttributeUniquer;
Traits = {mlir::triton::gpu::TritonGPU_AttrTrait::Trait, mlir::triton::gpu::DistributedEncodingTrait::Trait}]:
Assertion `succeeded(ConcreteT::verify(getDefaultDiagnosticEmitFn(ctx), args...))' failed.

PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.

Stack dump:
...
```

> So we don't need to fix this attribute's C++ code per se?

In case of internal error, failed code will be straight down the call stack.

Reply:
ok, nevermind, I'll change this is IR formatting, as long as the error string is unique it should be easy to grep it in source code anyway.